webknossos/webknossos/skeleton/graph.py

```diff
@@ -13,51 +13,143 @@
 Vector4 = Tuple[float, float, float, float]
 
 
-@attr.define(kw_only=True)
-class Graph:
+def _get_id(node_or_id: Union[Node, int]) -> int:
+    if isinstance(node_or_id, Node):
+        return node_or_id.id
+    else:
+        return node_or_id
+
+
+class _NodeDict(MutableMapping):
+    """Dict-like container for the `Graph` class below to handle both nodes and ids as keys.
+    Only when setting the value of a node for the first time the actual node must be passed.
+    This allows to keep a reference from all ids to the nodes, which can be looked up
+    using `get_node()`. Iterating over the keys yields the `Node` objects."""
+
+    def __init__(self) -> None:
+        self._id_to_attrs: Dict[int, Any] = {}
+        self._id_to_node: Dict[int, Node] = {}
+
+    def __getitem__(self, key: Union[Node, int]) -> Any:
+        return self._id_to_attrs[_get_id(key)]
+
+    def __setitem__(self, key: Union[Node, int], value: Dict) -> None:
+        key_id = _get_id(key)
+        if key_id not in self._id_to_node:
+            if isinstance(key, Node):
+                self._id_to_node[key_id] = key
+            else:
+                raise ValueError(
+                    f"Tried to add node {key}, which does not exist yet, to a graph."
+                    + "For insertion, the node must be an instance of the Node class."
+                )
+        self._id_to_attrs[key_id] = value
+
+    def __delitem__(self, key: Union[Node, int]) -> None:
+        del self._id_to_node[_get_id(key)]
+        del self._id_to_attrs[_get_id(key)]
+
+    def __iter__(self) -> Iterator[Node]:
+        return iter(self._id_to_node.values())
+
+    def __len__(self) -> int:
+        return len(self._id_to_attrs)
+
+    def get_node(self, id_: int) -> Node:
+        return self._id_to_node[id_]
+
+
+class _AdjDict(MutableMapping):
+    """Dict-like container for the `Graph` class below to handle both nodes and ids as keys.
+    Needs a reference to the _node attribute (of class `_NodeDict`) of the graph object
+    to get a reference from ids to nodes. Iterating over the keys yields the `Node` objects."""
+
+    def __init__(self, *, node_dict: _NodeDict) -> None:
+        self._id_to_attrs: Dict[int, Any] = {}
+        self._node_dict = node_dict
+
+    def __getitem__(self, key: Union[Node, int]) -> Any:
+        return self._id_to_attrs[_get_id(key)]
+
+    def __setitem__(self, key: Node, value: Dict) -> None:
```

Comment:
Good point, I simply missed the correct type annotation. I'd even skip the assertion, since `_AdjDict` doesn't store the id-to-node mapping and we don't know if it's inserted first into `_NodeDict` or `_AdjDict`.